src/main/java/MealDemo/service/MealPlanningServiceImpl.java

```diff
@@ -82,4 +80,84 @@ public void updateMeal(int mealId, Meal meal) {
         System.out.println("meal has been updated. Meal details \n" + existingMeal.toString());
     }
 
-}
\ No newline at end of file
+
+    @Override
+    public Map<String, List<Meal>> mealsByCategories() {
+        List<Meal> allMeals = getMeals();
+        Map<String, List<Meal>> mealsByCategoriesList = new HashMap<>();
+        for (Meal meal : allMeals) {
+            mealsByCategoriesList.computeIfAbsent(meal.getCategory(), k -> new ArrayList<>()).add(meal);
+        }
+
+        return mealsByCategoriesList;
+    }
+
+
+    private int randomNumberGenerator(int max) {
+        Random random = new Random();
+        return random.nextInt(max);
+    }
+
+
+    @Override
+    // generate 7 random meals per each day
+    public HashMap<DaysOfTheWeek, Meal> generateWeeklyMeals(HashMap<String, Integer> userInput) {
+        HashMap<DaysOfTheWeek, Meal> weeklyPlan = new HashMap<>();
+        Map<String, List<Meal>> mealsByCategories = mealsByCategories();
+
+        HashMap<String, Integer> request = new HashMap<>(userInput);
+        if (request.isEmpty()) {
+            request.put("Chicken", 2);
+            request.put("Fish", 1);
+            request.put("Meat", 2);
+            request.put("Vegetarian", 2);
+        }
+
+        for (DaysOfTheWeek day : DaysOfTheWeek.values()) {
+            String[] categories = request.keySet().toArray(new String[request.size()]);
+            String randomCategory = categories[new Random().nextInt(categories.length)];
+            int requestNumber = request.get(randomCategory);
+
+            List<Meal> meal = mealsByCategories.get(randomCategory);
+
+            Meal randomMeal = meal.get(randomNumberGenerator(meal.size()));
+            weeklyPlan.putIfAbsent(day, randomMeal);
+            requestNumber --;
+
+            if (!meal.isEmpty()) {
+                meal.remove(randomMeal);
+            }
+
+            if (requestNumber == 0) {
+                request.remove(randomCategory);
+            } else {
+                request.replace(randomCategory, requestNumber);
+            }
+        }
+
+        return weeklyPlan;
+    }
+
+    @Override
+    public HashMap<String, Integer> requestLog(HashMap<String, Integer> userInput) {
```

Comment:
This is not used by any controller (unless I missed it). Why do you need this method?

Reply:
You are correct. I started with this method to validate user input, what after writing tests, the validation moved to the controller